framework/src/multiapps/MultiApp.C

```diff
@@ -421,10 +429,40 @@ MultiApp::restore()
   if (_apps.empty())
     return;
 
+  if (_keep_solution_during_restore)
+  {
+    _end_solutions.resize(_my_num_apps);
+
+    for (unsigned int i = 0; i < _my_num_apps; i++)
+    {
+      _end_solutions[i] =
+          _apps[i]->getExecutioner()->feProblem().getNonlinearSystemBase().solution().clone();
+      auto & sub_multiapps =
+          _apps[i]->getExecutioner()->feProblem().getMultiAppWarehouse().getObjects();
+
+      for (auto & multi_app : sub_multiapps)
```

Comment:
So you spent three days on this fix, but you didn't add a single comment of what was the crux of the fix. It seems that it might be useful for others to know and perhaps will serve as a reminder for what and why you made that change. I assume that these two lines are the essential piece, right?

If I understand correctly, it's important to force keeping the solution on down the Multiapp chain. However, this opens a small inconsistency. I believe if I set this option in the sub app file explicitly one way or the other, it'll get overridden by the master? If I take the time to set that shouldn't I at least be notified that my option is being overridden? Depending on how for you want to go, you might even consider a warning (which can be turned into an error with the right CLI options).

Reply:
Yes, these lines are important. But the lines that I spent 90% time on are:

```

     /* We need to synchronize solution so that local_solution has the right values */
       _apps[i]->getExecutioner()->feProblem().getNonlinearSystemBase().update();
```

We have two versions of solution in moose: parallel vector and ghosted vector. In the original implementation, when we copy the latest solution to the restated apps, we did not sync the global parallel vector to the ghosted vector, and then the ghosted vector does not have the right solution. It is why `keep_solution_during_restore` did not at the first place.

It was really hard to track down this kind of issue. 

>If I understand correctly, it's important to force keeping the solution on down the Multiapp >chain. However, this opens a small inconsistency.

Right, I was thinking two options: 

(1) User set a flag in master, and then we force this option through all subapps. Easy for users to use?

(2) User set flags to all apps. In this case, user control everything. But it is easy to make a mistake.

I am fine  to take either way.

Reply:
> Right, I was thinking two options:
> 
> (1) User set a flag in master, and then we force this option through all subapps. Easy for users to use?
> 
> (2) User set flags to all apps. In this case, user control everything. But it is easy to make a mistake.
> 
> I am fine to take either way.

Which way you like @vincentlaboure ?

Reply:
> > Right, I was thinking two options:
> > (1) User set a flag in master, and then we force this option through all subapps. Easy for users to use?
> > (2) User set flags to all apps. In this case, user control everything. But it is easy to make a mistake.
> > I am fine to take either way.
> 
> Which way you like @vincentlaboure ?

If the two options are equivalent, I would tend to prefer the less-error-prone way (i.e. (1)) but both ways seem okay to me.

Reply:
I definitely think the first option is better. I just want to make sure that you give out the right errors when users try to set these options in the subapps and wonder why those options are silently ignored.